### PR TITLE
Coralogix processor first PR

### PR DESCRIPTION
**Pull request by galrose**
**Description:** <Describe what has changed.>
Adding a feature - Adding a feature to create templates (blueprints) from sql queries.
Currently specifically for postgresql and mysql queries.

**Link to tracking Issue:** #33090

**Testing:** <Describe what testing was performed and which tests were added.>
currently no tests, will be added in next PR

**Documentation:** <Describe the documentation added.>
Added documentation for possible configuration and the usecase of the processor

### CODE REVIEW

File: processor/coralogixprocessor/README.md
Comment (crobert-1) [inline processor/coralogixprocessor/README.md:20]: It looks like this feature is Coralogix-specific, can we potentially add a documentation link to make the context of this more clear? (Or if not a documentation link, more information in general to understand the purpose here)

File: processor/coralogixprocessor/README.md
Comment (crobert-1) [inline processor/coralogixprocessor/README.md:17]: I'm unfamiliar with Coralogix, but is this for a specific product or part of Coralogix? If so, should we add a link?

**Comment by crobert-1** [comment]
This PR should have a changelog entry. 👍

File: processor/coralogixprocessor/README.md
Comment (crobert-1) [inline processor/coralogixprocessor/README.md:30]: It would be good to include each config option, whether it's required, its default value, and a description.

File: processor/coralogixprocessor/metadata.yaml
Comment (crobert-1) [inline processor/coralogixprocessor/metadata.yaml:7]:
```suggestion
  distributions: []
```
We shouldn't mark it as being in `contrib` until it's included in `contrib` releases. Otherwise, it's easy for users to get confused and wonder why they can't use it with released `contrib` images.

File: processor/coralogixprocessor/config.go
Comment (crobert-1) [inline processor/coralogixprocessor/config.go:5]: Does this need to be public?

File: processor/coralogixprocessor/config.go
Comment (crobert-1) [inline processor/coralogixprocessor/config.go:16]: Do all of these need to be public? In general we want to have as few public members as possible.

**Review by crobert-1 — COMMENTED** [review]
Looks good overall. I've left a lot of comments on the README, the goal is to be able to flesh out UI/UX up front, before getting too far into the implementation weeds.

There are also a number of CI/CD failures caused by this PR, so please take a look and resolve when you're able. Let me know if you run into any issues there, happy to help!

File: processor/coralogixprocessor/README.md
Comment (galrose) [reply processor/coralogixprocessor/README.md:20]: We don't have docs for this feature yet, once we create the docs I will add them

File: processor/coralogixprocessor/README.md
Comment (galrose) [reply processor/coralogixprocessor/README.md]: [71a48f9](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33458/commits/71a48f9bc777d75420b43259c42b57756067bebf)

File: processor/coralogixprocessor/README.md
Comment (galrose) [reply processor/coralogixprocessor/README.md:39]: [71a48f9](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33458/commits/71a48f9bc777d75420b43259c42b57756067bebf)

File: processor/coralogixprocessor/README.md
Comment (galrose) [reply processor/coralogixprocessor/README.md]: will add link once docs are created

File: processor/coralogixprocessor/README.md
Comment (galrose) [reply processor/coralogixprocessor/README.md:56]: [71a48f9](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33458/commits/71a48f9bc777d75420b43259c42b57756067bebf)

File: processor/coralogixprocessor/README.md
Comment (galrose) [reply processor/coralogixprocessor/README.md]: [71a48f9](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33458/commits/71a48f9bc777d75420b43259c42b57756067bebf)

File: processor/coralogixprocessor/README.md
Comment (galrose) [reply processor/coralogixprocessor/README.md]: [71a48f9](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33458/commits/71a48f9bc777d75420b43259c42b57756067bebf)

I hope the added explanation is enough, if not maybe I should add an example

File: processor/coralogixprocessor/README.md
Comment (galrose) [reply processor/coralogixprocessor/README.md]: [71a48f9](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33458/commits/71a48f9bc777d75420b43259c42b57756067bebf)

File: processor/coralogixprocessor/span.go
Comment (galrose) [reply processor/coralogixprocessor/span.go:1]: [70e11a5](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33458/commits/70e11a571a3a04acfeb8ee53d885f725c78d8275)

File: processor/coralogixprocessor/factory.go
Comment (galrose) [reply processor/coralogixprocessor/factory.go]: [70e11a5](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33458/commits/70e11a571a3a04acfeb8ee53d885f725c78d8275)

File: processor/coralogixprocessor/config.go
Comment (galrose) [reply processor/coralogixprocessor/config.go]: [70e11a5](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33458/commits/70e11a571a3a04acfeb8ee53d885f725c78d8275)

File: processor/coralogixprocessor/config.go
Comment (galrose) [reply processor/coralogixprocessor/config.go]: [70e11a5](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33458/commits/70e11a571a3a04acfeb8ee53d885f725c78d8275)

File: processor/coralogixprocessor/config.go
Comment (galrose) [reply processor/coralogixprocessor/config.go]: [70e11a5](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33458/commits/70e11a571a3a04acfeb8ee53d885f725c78d8275)

File: processor/coralogixprocessor/README.md
Comment (crobert-1) [inline processor/coralogixprocessor/README.md:69]:
```suggestion
processors:
  coralogix:
    db_statement_blueprints:
      sampling: 
        enabled: true
        max_cache_size_mib: 1048576 #1GB
        max_cached_entries: 10000000 # 1 million
```
Do you think having the cache options as a subsection under `sampling` would make the functionality dependency more clear?

The main concern I'd have with my suggestion is if the cache ever needs to be used for something else. I'll defer to you to know future plans here.

File: processor/coralogixprocessor/config.go
Comment (crobert-1) [inline processor/coralogixprocessor/config.go:6]:
```suggestion
	maxCacheSizeMib int64 `mapstructure:"max_cache_size_mib"`
```
If you want the name to be a bit more concise.

File: processor/coralogixprocessor/config.go
Comment (crobert-1) [inline processor/coralogixprocessor/config.go:32]:
```suggestion
			return fmt.Errorf("max_cached_entries cannot be set if with_sampling is false")
```

File: processor/coralogixprocessor/config.go
Comment (crobert-1) [inline processor/coralogixprocessor/config.go:20]:
```suggestion
	if c.databaseBlueprintsConfig.withSampling {
```

File: processor/coralogixprocessor/config.go
Comment (crobert-1) [inline processor/coralogixprocessor/config.go:22]:
```suggestion
			return fmt.Errorf("max_cache_size_mib must be a positive integer")
```

File: processor/coralogixprocessor/config.go
Comment (crobert-1) [inline processor/coralogixprocessor/config.go:29]:
```suggestion
			return fmt.Errorf("max_cache_size_mib cannot be set if with_sampling is false")
```

File: processor/coralogixprocessor/README.md
Comment (crobert-1) [inline processor/coralogixprocessor/README.md:43]:
```suggestion
    * `max_cache_size_mib` (default: `1024`) The size of the cache in mebibytes to store seen blueprints hashes.
```
Default should be 1 GiB?

File: processor/coralogixprocessor/README.md
Comment (galrose) [reply processor/coralogixprocessor/README.md:42]: I will move the cache config under the sampling key

File: processor/coralogixprocessor/README.md
Comment (galrose) [reply processor/coralogixprocessor/README.md:44]: I will remove the max_cached_entries, because I store hashes I have a way to calculate the max entries from size of the cache

File: processor/coralogixprocessor/README.md
Comment (galrose) [reply processor/coralogixprocessor/README.md]: [5811865](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33458/commits/581186527bc4fbb9f8fc16a5c5c62d93117d695b)

File: processor/coralogixprocessor/config.go
Comment (galrose) [reply processor/coralogixprocessor/config.go]: [5811865](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33458/commits/581186527bc4fbb9f8fc16a5c5c62d93117d695b)

File: processor/coralogixprocessor/config.go
Comment (galrose) [reply processor/coralogixprocessor/config.go]: [5811865](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33458/commits/581186527bc4fbb9f8fc16a5c5c62d93117d695b)

File: processor/coralogixprocessor/config.go
Comment (galrose) [reply processor/coralogixprocessor/config.go]: [5811865](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33458/commits/581186527bc4fbb9f8fc16a5c5c62d93117d695b)

File: processor/coralogixprocessor/README.md
Comment (crobert-1) [inline processor/coralogixprocessor/README.md:61]: Do we want to add a bool `enabled` config option here to show if caching is enabled or not? Without it

**Comment by galrose** [comment]
@crobert-1 can you please help me understand why I don't pass the CI?
I am running all the scripts from the first PR section and they all look fine locally

**Comment by crobert-1** [comment]
> @crobert-1 can you please help me understand why I don't pass the CI? I am running all the scripts from the first PR section and they all look fine locally

You'll have to look through the different failed actions individually to see why they're failing. From what I can tell they're related to this PR.

One [example](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/9561217316/job/26375214337?pr=33458):
```
license header checking failed:
./config.go
./config.go
./config.go
./config.go
./factory.go
./factory.go
./factory.go
./factory.go
./span.go
./span.go
./span.go
./span.go
make[2]: *** [../../Makefile.Common:174: checklicense] Error 1
make[1]: *** [Makefile:179: processor/coralogixprocessor] Error 2
```
This means these files don't have the proper license header. If you look in `Makefile.Common` you can see what this check is doing. You need to run `make addlicense` to resolve this.

For [this failure](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/9561217316/job/26375204100?pr=33458):
```
FAIL: "processor/coralogixprocessor" not included in CODEOWNERS
```
The command `make update-codeowners` needs to be run, as described in the [CONTRIBUTING doc](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-new-components).

There are more failures that need investigated, but let me know if there's anything you don't understand 👍

File: processor/coralogixprocessor/README.md
Comment (crobert-1) [reply processor/coralogixprocessor/README.md:61]: The question that I'm getting at is with the current config, how would a user disable sampling? It's not obvious from current configuration options and the README.

**Review by MovieStoreGuy — COMMENTED** [review]
All the changes seem fine, I agree with @crobert-1 that having some more clarity on disabling sampling would be nice.

File: processor/coralogixprocessor/README.md
Comment (galrose) [reply processor/coralogixprocessor/README.md:61]: Im not sure what the best way to describe it in the README but just not adding the key means not enabling it, similar to other processors, do you have a good idea how I should explain it?

**Comment by galrose** [comment]
Hey @crobert-1, thanks for adding the add license to the docs 😄 
Im getting an error when running the make update-codeowners it doesn't matter what I put in the meta.yaml

```
cd cmd/githubgen && go install .
githubgen
2024/06/20 12:13:21 codeowners are not members: BinaryFissionGames, Frapschen, JonathanWamsley, MovieStoreGuy, Mrod1598, RichieSams, StefanKurek, VihasMakwana, adrielp, ag-ramachandran, akats7, alexvanboxel, alxbl, ankitpatel96, antonblock, armstrmi, atingchen, atoulme, avadhut123pisal, bertysentry, braydonk, bryan-aguilar, carsonip, chrroberts-pure, codeboten, cparkins, cpheps, damemi, dashpole, dehaansa, dgoscn, eedorenko, erichsueh3, evan-bradley, fatsheep9146, fchikwekwe, fjslkfjdsklfj, gjanco, greatestusername, hickeyma, hughesjj, jacobmarble, jamesmoessis, jmacd, jsuereth, khyatigandhi6, kkujawa-sumo, kongluoxing, liustanley, mapno, mar4uk, marctc, martin-majlis-s1, mat-rumian, matej-g, mcardy, michalpristas, moh-osman3, mrsillydog, nslaughter, pavankrish123, pcwiese, pjanotti, portertech, povilasv, psx95, pxaws, qiansheng91, rapphil, rmfitzpatrick, rnishtala-sumo, rogercoll, samiura, schmikei, sh0rez, shabicheng, shalper2, sinkingpoint, smithclay, splunkericl, srprash, swiatekm-sumo, tamir-michaeli, tomaz-s1, wangzlei, ycombinator, yotamloe, zdaratom-s1
make: *** [gengithub] Error 1
```

I think these are the actual code owners but adding them to the list instead of me still doesn't change the error. 😅

File: processor/coralogixprocessor/README.md
Comment (galrose) [reply processor/coralogixprocessor/README.md:63]: I added a comment about it in this [commit](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33458/commits/d68c182a69d167fc0fbccd2efb86c6a07c7fc8d6), do you think this is enough?

Comment (crobert-1) [reply processor/coralogixprocessor/README.md:63]: A component's configuration has no way of knowing if a key is included in the configuration, as far as I know. The collector uses [mapstructure](https://github.com/go-viper/mapstructure) to unmarshal YAML files into configuration options. If there's no value for a key (`sampling` in this case), there's nothing to unmarshal. You'd need a value to be stored to know if a configuration option was specified or not.

If I've missed something please let me know. (You mention other processors do this, so if my explanation doesn't make sense, feel free to give an example.)

Comment (galrose) [reply processor/coralogixprocessor/README.md]: [2dd6c34](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33458/commits/2dd6c3417678af6d05b2803b5e69192f8535851c)

added the enabled key

**Comment by galrose** [comment]
Hey @crobert-1, is there a chance you can take a look at why update-codeowners isnt working for me?

**Comment by crobert-1** [comment]
> Hey @crobert-1, is there a chance you can take a look at why update-codeowners isnt working for me?

It looks like you may need to add yourself to [this file](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/cmd/githubgen/allowlist.txt).

**Comment by crobert-1** [comment]
It also looks like you need to run `make gotidy` to fix CI/CD failures.

File: .github/CODEOWNERS
Comment (crobert-1) [inline .github/CODEOWNERS:164]: This should also include me as a code owner.

File: processor/coralogixprocessor/README.md
Comment (crobert-1) [inline processor/coralogixprocessor/README.md:24]:
```suggestion
The blueprint is also hashed to be able to be used with the spanmetrics connector.
```

File: processor/coralogixprocessor/README.md
Comment (crobert-1) [inline processor/coralogixprocessor/README.md:31]:
```suggestion
* `db.statement.blueprint` contains the blueprinted version of the statement, we require them to be sent to Coralogix to
```

File: processor/coralogixprocessor/README.md
Comment (crobert-1) [inline processor/coralogixprocessor/README.md:48]:
```suggestion
to only send new blueprints to Coralogix.
```

File: processor/coralogixprocessor/README.md
Comment (crobert-1) [inline processor/coralogixprocessor/README.md:49]:
```suggestion
If sampling is not enabled it won't cache anything and the `sampling.priority` attribute won't be added.
```

File: processor/coralogixprocessor/config.go
Comment (crobert-1) [inline processor/coralogixprocessor/config.go:8]:
```suggestion
type samplingConfig struct {
```
Please update references as well

File: processor/coralogixprocessor/config.go
Comment (crobert-1) [inline processor/coralogixprocessor/config.go:26]:
```suggestion
		return fmt.Errorf("max_cache_size_mib can only be defined when sampling is enabled")
```

File: processor/coralogixprocessor/config.go
Comment (galrose) [reply processor/coralogixprocessor/config.go]: [26e7c79](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33458/commits/26e7c791ce217ddffeca2e7b3f5e363dfebd96a8)

**Review by crobert-1 — APPROVED** [review]

**Comment by galrose** [comment]
@atoulme I see I need your approval as well for the code owners 😄

**Review by atoulme — APPROVED** [review]

**Comment by crobert-1** [comment]
Looks like there are some more check failures:
```
go.mod/go.sum deps changes detected, please run "make gotidy" and commit the changes in this PR.
```
_There may be more than that, I didn't check thoroughly._

File: .github/CODEOWNERS
Comment (songy23) [inline .github/CODEOWNERS:168]: This broke a CI in mainline: https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/10317079346/job/28560709557#step:16:11

> codeowners are not members: eyalatz, roycald245

Comment (roycald245) [reply .github/CODEOWNERS:168]: how do we fix this? is this because we are missing in `cmd/githubgen/allowlist.txt`?

**Pull request merged**
